File: bindings/python/protocoll/state.py
"""State-related types and error handling."""

from enum import IntEnum


class CrdtType(IntEnum):
    """CRDT type for a state region."""
    LWW_REGISTER = 0
    G_COUNTER = 1
    PN_COUNTER = 2
    OR_SET = 3


class Reliability(IntEnum):
    """Reliability level for state propagation."""
    RELIABLE = 0
    BEST_EFFORT = 1


class ResolutionTier(IntEnum):
    """Resolution tier controlling how much detail is synced for a node."""
    FULL = 0
    NORMAL = 1
    COARSE = 2
    METADATA = 3


class ProtocollError(Exception):
    """Base exception for protocoll errors."""
    pass


class InvalidError(ProtocollError):
    """Invalid argument."""
    pass


class NotFoundError(ProtocollError):
    """Path not declared."""
    pass


class NoConnectionError(ProtocollError):
    """Not connected."""
    pass


class TimeoutError(ProtocollError):
    """Operation timed out."""
    pass


class CryptoError(ProtocollError):
    """Signature verification failed."""
    pass


_ERROR_MAP = {
    -1: InvalidError,
    -2: NotFoundError,
    -3: NoConnectionError,
    -4: TimeoutError,
    -5: CryptoError,
}


def _check_error(code: int):
    """Convert a C API error code to a Python exception."""
    if code == 0:
        return
    exc_class = _ERROR_MAP.get(code, ProtocollError)
    raise exc_class(f"protocoll error code: {code}")

File: bindings/rust/protocoll/src/state.rs
//! State-related types for CRDT and reliability configuration.

/// CRDT type for a state region.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CrdtType {
    /// Last-Writer-Wins Register.
    LwwRegister,
    /// Grow-only Counter.
    GCounter,
    /// Positive-Negative Counter.
    PnCounter,
    /// Observed-Remove Set.
    OrSet,
}

impl CrdtType {
    pub(crate) fn to_raw(self) -> protocoll_sys::PcolCrdtType {
        match self {
            CrdtType::LwwRegister => protocoll_sys::PcolCrdtType_PCOL_LWW_REGISTER,
            CrdtType::GCounter => protocoll_sys::PcolCrdtType_PCOL_G_COUNTER,
            CrdtType::PnCounter => protocoll_sys::PcolCrdtType_PCOL_PN_COUNTER,
            CrdtType::OrSet => protocoll_sys::PcolCrdtType_PCOL_OR_SET,
        }
    }
}

/// Reliability level for state propagation.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Reliability {
    /// Full ACK/retransmit.
    Reliable,
    /// Retransmit once, then drop.
    BestEffort,
}

impl Reliability {
    pub(crate) fn to_raw(self) -> protocoll_sys::PcolReliability {
        match self {
            Reliability::Reliable => protocoll_sys::PcolReliability_PCOL_RELIABLE,
            Reliability::BestEffort => protocoll_sys::PcolReliability_PCOL_BEST_EFFORT,
        }
    }
}

/// Resolution tier controlling how much detail is synced to a remote node.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ResolutionTier {
    /// Full resolution — all deltas.
    Full,
    /// Normal resolution.
    Normal,
    /// Coarse resolution — reduced update frequency.
    Coarse,
    /// Metadata only — no payload data.
    Metadata,
}

impl ResolutionTier {
    pub(crate) fn to_raw(self) -> protocoll_sys::PcolResolutionTier {
        match self {
            ResolutionTier::Full => protocoll_sys::PcolResolutionTier_PCOL_RESOLUTION_FULL,
            ResolutionTier::Normal => protocoll_sys::PcolResolutionTier_PCOL_RESOLUTION_NORMAL,
            ResolutionTier::Coarse => protocoll_sys::PcolResolutionTier_PCOL_RESOLUTION_COARSE,
            ResolutionTier::Metadata => protocoll_sys::PcolResolutionTier_PCOL_RESOLUTION_METADATA,
        }
    }
}
